crates/miden-tx/src/executor/notes_checker.rs

```diff
@@ -23,17 +26,96 @@ where
         NoteConsumptionChecker(tx_executor)
     }
 
-    /// Checks whether the provided input notes could be consumed by the provided account by
-    /// executing the transaction.
+    /// Checks whether some set of the provided input notes could be consumed by the provided
+    /// account by executing the transaction with varying combination of notes.
+    ///
+    /// This function attempts to find the maximum set of notes that can be successfully executed
+    /// together by the target account.
+    ///
+    /// If some notes succeed but others fail, the failed notes are removed from the candidate set
+    /// and the remaining notes (successful + unattempted) are retried in the next iteration. This
+    /// process continues until either all remaining notes succeed or no notes can be successfully
+    /// executed
+    ///
+    /// # Example Execution Flow
+    ///
+    /// Given notes A, B, C, D, E:
+    /// - Try [A, B, C, D, E] → A, B succeed, C fails → Remove C, try again.
+    /// - Try [A, B, D, E] → A, B, D succeed, E fails → Remove E, try again.
+    /// - Try [A, B, D] → All succeed → Return successful=[A, B, D], failed=[C, E].
+    ///
+    /// # Returns
+    ///
+    /// Returns [`NoteConsumptionInfo`] containing:
+    /// - `successful`: Notes that can be consumed together by the account.
+    /// - `failed`: Notes that failed during execution attempts.
     pub async fn check_notes_consumability(
         &self,
         target_account_id: AccountId,
         block_ref: BlockNumber,
         input_notes: InputNotes<InputNote>,
         tx_args: TransactionArgs,
     ) -> Result<NoteConsumptionInfo, TransactionExecutorError> {
-        self.0
-            .try_execute_notes(target_account_id, block_ref, input_notes, tx_args)
+        // Ensure well-known notes are ordered first.
+        let mut notes = input_notes.into_vec();
+        notes.sort_by_key(|note| WellKnownNote::from_note(note.note()).is_none());
```

Comment:
Do we need a stable sort here? We just care about that the well-known notes are at the front, but we don't care about the exact order, right? If so, we could use `sort_unstable_by_key` which is a bit more efficient.